### PR TITLE
C++ improvements

### PR DESCRIPTION
**Pull request by bcebere**
## Description
C++ interface updates.
 - [x] Use shared_ptr for tensor API
 - [x] Add abstract EncryptedTensor interface, as a baseline.
 - [x] Add abstract EncryptedVector interface, that specializes EncryptedTensor for vectors.
 - [x] Update header docs.

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

**Comment by youben11** [comment]
Really excited about this PR!

**Review by youben11 — COMMENTED** [review]
Q: Why are we switching to shared_ptr for the tensor objects?

**Comment by bcebere** [comment]
In order to move to abstract tensor classes, I have to move to pointer-based management.

**Review by youben11 — COMMENTED** [review]
Starting to look amazing!
I have 2 general comments:
1. I see that BFV is now operating on vectors of double, which might be unclear for the user, as he expect BFV to work on integers, how is the users supposed to be blocked in an attempt to pass real numbers, will he know that it only uses integers?
2. I know that we have been using vectors extensively until now, but I expect the more general parent class (EncryptedTensor) to not implement tensor specific operations. Another EncryptedVector class can inherit from the EncryptedTensor and implement the recurrent methods from CKKS and BFV Vectors. This will allow to develop the new tensor type that can implement his own logic, which would be really different from the ones implemented for vectors.

**Review by youben11 — CHANGES_REQUESTED** [review]
Awesome Bogdan! I really liked that all those changes didn't require any actual changes to the python tests, which means that the API is still the same, great job!

I left some comments as open discussions, to better choose the design for the upcoming changes relating to the new tensor type.

File: tenseal/binding.cpp [hunk 1]
```diff
@@ -41,46 +37,81 @@ PYBIND11_MODULE(_tenseal_cpp, m) {
...
-    py::class_<BFVVector>(m, "BFVVector")
-        .def(py::init<shared_ptr<TenSEALContext> &, vector<int64_t>>())
-        .def(py::init<shared_ptr<TenSEALContext> &, const std::string &>())
-        .def("size", &BFVVector::size)
+    py::class_<EncryptedVector<int64_t>,
+               std::shared_ptr<EncryptedVector<int64_t>>>(
+        m, "EncryptedVectorInt64", py::module_local());
+    py::class_<EncryptedVector<double>,
+               std::shared_ptr<EncryptedVector<double>>>(
+        m, "EncryptedVectorDouble", py::module_local());
```
[the remark this answers is not in the document]
Comment (bcebere) [reply tenseal/binding.cpp:45]: pybind requires them, because the operations return EncryptedVector instances.

On a second thought, I could change the operations to always return an instance from the current class

File: tenseal/cpp/tensors/encrypted_tensor.h [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+using namespace seal;
+using namespace std;
+
+/**
+ * EncryptedTensor<plain_t, plain_container_t, encrypted_t> interface - A
+ *generic API for encrypted tensor operations.
+ * @param plain_t: root plaintext datatype for representing data(double, int64
+ *etc).
+ * @param plain_container_t: complete plaintext representation(for example,
+ *vector<double>).
```
[the remark this answers is not in the document]
Comment (bcebere) [reply tenseal/cpp/tensors/encrypted_tensor.h:21]: I can add a PlainTensor class that wraps the plain_container_t, and will have constructors for any data type we want to support. 

It would help me to know what abstract operations should I add to the Encrypted/PlainTensor classes.

So far, there should be:
 - shape/reshape?
 - broadcast?

should I add tensorflow operations to the interface, as a reference?

File: tenseal/cpp/tensors/encrypted_tensor.h [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    /**
+     * Polynomial evaluation with `this` as variable.
+     * p(x) = coefficients[0] + coefficients[1] * x + ... + coefficients[i] *
+     *x^i
+     **/
+    encrypted_t polyval(const plain_container_t& coefficients) const {
+        return this->copy()->polyval_inplace(coefficients);
+    };
+    virtual encrypted_t polyval_inplace(
+        const plain_container_t& coefficients) = 0;
```
[the remark this answers is not in the document]
Comment (bcebere) [reply tenseal/cpp/tensors/encrypted_tensor.h:128]: you are right, my bad

File: tenseal/cpp/tensors/encrypted_tensor.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (youben11) [reply tenseal/cpp/tensors/encrypted_tensor.h:21]: If we add a single class, how could we distinguish between vectors and multi-dimensional tensors? using the shape only? So operations like add_plain would need to check the shape and see if they can operate with that specific tensor.

For the operations, I guess we can start with just a small subset that is required now, and add as we need them.

[the remark this answers is not in the document]
Comment (bcebere) [reply tenseal/cpp/tensors/encrypted_tensor.h:21]: So, in almost every implementation, tensors have the following properties:
 - **an 1-D vector** for internal representation, of size D * W * H
 - **a size/shape tuple**, with the current view of the 1-D vector, for 3D let's say (D, H, W)
 - **a stride stuple** meaning how much you jump for every dimension in the 1D representation, in the previous example (H * W, W, 1).
 - **the data type**, that plain_t, int64_t/double in our cases.
 - **layout of the data**, if we want to use a map instead of vector, no idea.
 - **device**, if we support CPU/CUDA etc.

right? 
we do not store the data reshaped in memory, we use those strides and shape to jump in the 1D representation.
This is taken from the PyTorch internals http://blog.ezyang.com/2019/05/pytorch-internals/


So, I should add a plaintext tensor with those properties.
And the encrypted one should have similar properties, to handle the strides/shapes/views.

In the current implementation, the encrypted tensor now has a broadcast operation.

Does this make sense?

[the remark this answers is not in the document]
Comment (youben11) [reply tenseal/cpp/tensors/encrypted_tensor.h:21]: Yeah absolutely, this is really great, and that's what I was also planning to implement for the arbitrary dimension tensor. It seems like all tensor library are using this representation, and it's way more efficient when we need to reshape, as only those steps need to be updated instead of moving data around.

[the remark this answers is not in the document]
Comment (bcebere) [reply tenseal/cpp/tensors/encrypted_tensor.h]: I added a PlainTensor class to handle the vector/matrix parameters for TenSEAL.

It's nice that we can easily make it compatible with PyTorch/Tensorflow tensors as well.
Let me know what you think, please.

File: tenseal/cpp/tensors/encrypted_vector.h [hunk 4]
```diff
@@ -69,78 +68,229 @@ class EncryptedVector
         return this->copy()->replicate_first_slot_inplace(n);
     }
     virtual encrypted_t replicate_first_slot_inplace(size_t n) = 0;
-    void broadcast_or_throw(encrypted_t other) {
-        if (this->size() == other->size()) return;
+    /**
+     * Adjust two vectors to match sizes.
+     * @return the right operand, in case it was copied and altered.
+     **/
+    encrypted_t broadcast_or_throw(encrypted_t other) {
```
Comment (youben11) [inline tenseal/cpp/tensors/encrypted_vector.h:75]: Would it help to pass `other` by reference since it's not modified?

**Review by youben11 — COMMENTED** [review]
I just left small comments then I think it's ready to merge!

File: tenseal/cpp/tensors/encrypted_vector.h (hunk 4 above)
Comment (youben11) [reply tenseal/cpp/tensors/encrypted_vector.h:75]: Ah sorry, I guess encrypted_t is already a pointer

File: tenseal/binding.cpp [hunk 5]
```diff
@@ -50,46 +50,74 @@ PYBIND11_MODULE(_tenseal_cpp, m) {
...
         .def("add", &BFVVector::add)
         .def("add_", &BFVVector::add_inplace)
-        .def("add_plain", py::overload_cast<const vector<int64_t> &>(
-                              &BFVVector::add_plain, py::const_))
-        .def("add_plain_", py::overload_cast<const vector<int64_t> &>(
-                               &BFVVector::add_plain_inplace))
+        .def("add_plain",
+             [](shared_ptr<BFVVector> obj, const vector<int64_t> &other) {
+                 return obj->add_plain(other);
+             })
```
[the remark this answers is not in the document]
Comment (bcebere) [reply tenseal/binding.cpp:65]: The operations now require a PlainTensor parameter and pybind doesn't know how to cast python vectors to PlainTensor, although PlainTensor has cast operator for std::vector.
Or I do not know how to indicate the cast to pybind.

File: tenseal/cpp/tensors/plain_tensor.h [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+   private:
+    vector<plain_t> _data;
+    vector<size_t> _shape;
+    vector<size_t> _strides;
+};
+
+}  // namespace tenseal
+
+#endif
```
[the remark this answers is not in the document]
Comment (bcebere) [reply tenseal/cpp/tensors/plain_tensor.h:146]: Thank you!

File: tenseal/cpp/tensors/encrypted_vector.h (hunk 4 above)
Comment (bcebere) [reply tenseal/cpp/tensors/encrypted_vector.h:75]: Yeah, I should use const reference everywhere, in case encrypted_t is not a pointer

File: tenseal/binding.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (bcebere) [reply tenseal/binding.cpp:65]: If you have any suggestion here, I can try it. Not sure what type uses pybind internally for python vector.

But it could be helpful if we expose similar overloads for PyTorch or Tensorflow tensors, right now doesn't seem that useful, true.

[the remark this answers is not in the document]
Comment (youben11) [reply tenseal/binding.cpp:65]: As far as I know, pybind11 can converts python list into vectors only, but I think we can add custom type casters so that pybind11 knows how to do it.
I only read about it, but it's supposed to work. With this I guess we can only provide custom casters for different types (python lists, torch tensors...) without the need to wrap each function.

[the remark this answers is not in the document]
Comment (bcebere) [reply tenseal/binding.cpp:65]: I think I need to define pybind11::class_ for PlainTensor instantions to make it work, I will try that.

[the remark this answers is not in the document]
Comment (youben11) [reply tenseal/binding.cpp:65]: I'm also up to leave it for another PR. Just let me know

[the remark this answers is not in the document]
Comment (bcebere) [reply tenseal/binding.cpp:86]: I am unable to make it work.

We need to implement a [PYBIND11_TYPE_CASTER](https://pybind11.readthedocs.io/en/stable/advanced/cast/custom.html) to make PlainTensor compatible with python lists, but I need to research this a bit more.

I'd leave it for another PR.

**Review by youben11 — APPROVED** [review]
